### PR TITLE
Import Session : Removed issues for Windows and some other code adjustments 

### PR DESCRIPTION
**Pull request by Jiyone**
- Management of "/" and "\" was broken on Windows. Escape of $ using "\$" removed: "$" will be written if it's not followed by a "(". Else, it's a $(thing).
- "/" are replaced by the correct directory separator on Windows.
- The GUI's folder separators between the different fields in Import are correct on Windows
- Callback added to jobcode field, so test_path updates after editing it;
- The "test path" building code is simplified
- It doesn't create folders each time it refreshes anymore.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -27,3 +27,4 @@ AppDir
 appimage-build
 linuxdeploy-plugin-gtk.sh
 src/tests/integration/*/*
+src/control/jobs/import.md
```
Comment (aurelienpierre) [inline .gitignore:30]: You could let the Markdown file here, even if it's not translated… Pending a real dev doc

File: src/common/import.c [hunk 2]
```diff
@@ -116,6 +117,8 @@ typedef struct dt_lib_import_t
 
   dt_pthread_mutex_t lock;
 
+  char *path_file;
```
Comment (aurelienpierre) [inline src/common/import.c:120]: Note: check that the `***_cleanup()` function frees `*path_file` whhen the `struct dt_import_t` is freed.

File: src/common/import.c [hunk 3]
```diff
@@ -560,49 +564,67 @@ static void _set_help_string(dt_lib_import_t *d, gboolean copy)
...
+                                };
+
+    dt_variables_params_t *params;
+    dt_variables_params_init(&params);
+
+    params->filename = g_strdup(file.data);
+    params->sequence = 1;
+    params->jobcode = g_strdup(data.jobcode);
+    
+    const gchar *fake_path = dt_build_filename_from_pattern(params, &data);
```
Comment (aurelienpierre) [inline src/common/import.c:620]: `*fake_path` needs to be freed before exiting the function

File: src/common/variables.c [hunk 4]
```diff
@@ -185,16 +200,11 @@ static void _init_expansion(dt_variables_params_t *params, gboolean iterate)
...
-    if(params->img == NULL) dt_image_cache_read_release(darktable.image_cache, img);
-  }
-  else if(params->data->exif_time[0])
-  {
-    params->data->datetime = dt_datetime_exif_to_gdatetime(params->data->exif_time, darktable.utc_tz);
-    if(params->data->datetime)
-      params->data->have_exif_dt = TRUE;
   }
+  // FIXME: do we have a deadlock if params->img != NULL ??
+  if(params->img == NULL) dt_image_cache_read_release(darktable.image_cache, img);
```
Comment (aurelienpierre) [inline src/common/variables.c:205]: Warning: `dt_image_cache_read_release` needs to be called only if `dt_image_cache_get` was called. Probably add a boolean variable in the `else if(params->imgid)` branch where the `...cache_get()` function is called, then only in that case release the cache.

File: src/common/variables.c [hunk 5]
```diff
@@ -185,16 +200,11 @@ static void _init_expansion(dt_variables_params_t *params, gboolean iterate)
...
+  
+//TODO: Check if img is freed when alloced from here.
```
Comment (aurelienpierre) [inline src/common/variables.c:207]: In `_cleanup_expansion()` please add 

```C
if (params->img)
    free(params->img);
```

There might also be an internal function to properly cleanup `(dt_image_t *)img` case some internal buffer needs deallocation first.

File: src/control/jobs/control_jobs.c [hunk 6]
```diff
@@ -2211,82 +2082,263 @@ static GList *_apply_lua_filter(GList *images)
...
+  }
+  return FALSE;
+}
 
-#ifdef USE_LUA
-  if(!data->session)
+gchar *_path_cleanup(gchar *path)
+{
+  const gchar *res = dt_cleanup_separators(path);
+  return dt_util_remove_whitespace(res);
```
Comment (aurelienpierre) [inline src/control/jobs/control_jobs.c:2098]: This is dangerous here. Strings can't be modified in-place, so you need to re-alloc new buffers to write the modified content. Meaning you need to free the old buffers if you are not going to reuse them. The pointer is only the address of the first element in the buffer. If you recycle the pointer `*path`, you are only redirecting it to the newly-allocated copies, but the original buffers still exist in memory. Meaning here you need to deal with `char *in` and `char *out` variables, and manually free the ones you don't need.

File: src/control/jobs/control_jobs.c [hunk 7]
```diff
@@ -2211,82 +2082,263 @@ static GList *_apply_lua_filter(GList *images)
...
+}
+
+gchar *dt_build_filename_from_pattern(dt_variables_params_t *params, dt_control_import_t *data)
+{
+  gchar *file_expand = dt_variables_expand(params, data->target_file_pattern, FALSE);
+  gchar *path_expand = dt_variables_expand(params, data->target_subfolder_pattern, FALSE);
+
+  // remove this if we decide to do the correction on user's settings directly 
+  file_expand = _path_cleanup(file_expand);
```
Comment (aurelienpierre) [inline src/control/jobs/control_jobs.c:2107]: Same as previous comment, use `char *in` and `char *out` buffer logic with manual freeing of unused buffers, reassigning pointers doesn't deallocate the buffers.

File: src/control/jobs/control_jobs.c [hunk 8]
```diff
@@ -2211,82 +2082,263 @@ static GList *_apply_lua_filter(GList *images)
...
+    if(first >= 'A' && first <= 'Z' && data->target_dir[1] == ':') // path format is <drive letter>:\path\to\file
+      data->target_dir[0] = first;                                 // drive letter in uppercase looks nicer
   }
 #endif
 
-  GList *t = params->index;
-  const guint total = g_list_length(t);
-  snprintf(message, sizeof(message), ngettext("importing %d image", "importing %d images", total), total);
-  dt_control_job_set_progress_message(job, message);
+  data->target_dir = dt_util_normalize_path(data->target_dir);
```
Comment (aurelienpierre) [inline src/control/jobs/control_jobs.c:2124]: Ensure `normalize_path` frees the original buffer

File: src/control/jobs/control_jobs.c [hunk 9]
```diff
@@ -2211,82 +2082,263 @@ static GList *_apply_lua_filter(GList *images)
...
+}
+
+// returns TRUE if success
+gboolean _copy_file(const char *filename, const char *dest_file_path)
+{
+  GFile *in = g_file_new_for_path(filename);
+  GFile *out = g_file_new_for_path(dest_file_path);
+
+  gboolean res = g_file_copy(in, out, G_FILE_COPY_NONE, 0, 0, 0, NULL);
+  res ? fprintf(stdout, "File copied successfully.\n")
```
Comment (aurelienpierre) [inline src/control/jobs/control_jobs.c:2162]: Use an `if(res) / else` statement here. Inline checks are reserved for variable assignation like `output = res ? something : something_else`.

File: src/control/jobs/control_jobs.c [hunk 10]
```diff
@@ -2321,36 +2372,21 @@ static dt_job_t *_control_import_job_create(GList *imgs, const char *datetime_ov
...
+  memcpy(params->data, &data, sizeof(dt_control_import_t));
+  fprintf(stdout, "END Job create.\n");
   return job;
 }
 
-void dt_control_import(GList *imgs, const char *datetime_override, const gboolean inplace)
+/**
+ * @brief Process a list of images to import with or without copying the files on an arbitrary hard-drive.
+ */
+void dt_control_import(dt_control_import_t data)
```
Comment (aurelienpierre) [inline src/control/jobs/control_jobs.c:2383]: This is weird. `dt_control_import_t` is a `struct`, it should be passed by pointer. That might explain why the last filmroll used wasn't properly written. I'm surprised it works at all.

**Review by aurelienpierre — CHANGES_REQUESTED** [review]
Please run `git fetch upstream` then `git rebase upstream master` (replace `upstream` with the actual name of this repo on your local Git).

I reviewed everything but the actual importing sequence (yet).

File: src/common/variables.c (hunk 4 above)
Comment (Jiyone) [reply src/common/variables.c:205]: Does that means I remove `params->img == NULL` and replace it with the true boolean value ?

**Pull request closed without merging**